src/scheduler.rs

```diff
@@ -0,0 +1,173 @@
+use std::collections::HashMap;
+use std::future::Future;
+use std::sync::{Arc, RwLock};
+use tokio::select;
+use tokio::time::{Duration, sleep, Instant};
+use tokio_util::sync::CancellationToken;
+
+type TaskId = u64;
+type TaskData = (CancellationToken, Instant);
+
+#[derive(Clone)]
+pub struct Scheduler {
+    tasks: Arc<RwLock<HashMap<TaskId, TaskData>>>,
+    task_duration: Duration,
+    start_token: CancellationToken,
+}
```

Comment:
As it's some kind of custom scheduler with your own logic, it would be awesome to add docs for other devs on this module and its members to ease the review process.